### DISCUSSION TITLE
Full quadrature encoder - 4x pulses per revolution resolution

### DISCUSSION THREAD

**Discussion by ilmarietto123** [kachurovskiy/nanoels#235] [General]
Hi,
I bought a 900 pulses per revolution encoder, thinking to achieve 0.1° resolution but I was surprised to read 0.2° on display.
I digged into the code and found out it uses half quadrature. 
I tried to implement a full quadrature and it works.
@kachurovskiy, is there some drawback that I'm missing?

Here are the modifications I did to the code:

[Line 6](https://github.com/kachurovskiy/nanoels/blob/8cc4b13d93870869a071ace68bb6aeda617e1e79/h4/h4.ino#L6C1-L6C106)
```
const int ENCODER_PPR = 900; // 600 step spindle optical rotary encoder. Fractional values not supported.
```
Use the specs value of your encoder multiplied by the gear ratio you have. In my case I have a 900ppr 1:1, so 900.

[Line 69](https://github.com/kachurovskiy/nanoels/blob/8cc4b13d93870869a071ace68bb6aeda617e1e79/h4/h4.ino#L69)
```
const int ENCODER_STEPS_INT = ENCODER_PPR * 4; // Number of encoder impulses PCNT counts per revolution of the spindle
```
Change from "* 2" to "* 4".

[Lines from 1588 to 1594](https://github.com/kachurovskiy/nanoels/blob/8cc4b13d93870869a071ace68bb6aeda617e1e79/h4/h4.ino#L1588-L1594)
```
  pcntConfig.pos_mode = PCNT_COUNT_DEC;       // changed
  pcntConfig.neg_mode = PCNT_COUNT_INC;       // changed
  pcntConfig.lctrl_mode = PCNT_MODE_REVERSE;
  pcntConfig.hctrl_mode = PCNT_MODE_KEEP;
  pcntConfig.counter_h_lim = PCNT_LIM;
  pcntConfig.counter_l_lim = -PCNT_LIM;
  pcnt_unit_config(&pcntConfig);
  pcntConfig.pulse_gpio_num = gpioB;          // added
  pcntConfig.ctrl_gpio_num = gpioA;           // added
  pcntConfig.channel = PCNT_CHANNEL_1;        // added
  pcntConfig.pos_mode = PCNT_COUNT_INC;       // added
  pcntConfig.neg_mode = PCNT_COUNT_DEC;       // added
  pcnt_unit_config(&pcntConfig);              // added
```
Replace this part of code or modify and add.

@kachurovskiy, @kotovasia5120: I saw in the last Pull request that you switched to "driver/pulse_cnt.h".
I tested the modifications except i2c display and the full quadrature seems to work too.
I don't post it here because it can make confusion.

**Comment by kachurovskiy** [comment]
Hi @ilmarietto123, great find, thanks for sharing! Hopefully I can incorporate this in one of the future software versions for H4 and H5.
